{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 167,
   "id": "c220f94f-a42c-48a5-91b4-995ff9710f19",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "622820e9-f767-4053-ba09-83b2244b828d",
   "metadata": {},
   "source": [
    "1. Create a small class with a constructor and several methods."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3db2cb53-6a99-4849-b358-90b86edf8743",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Dogs:\n",
    "    def __init__(self, age = 0, weight = 0, sex = 'undefined', breed = 'mongrel', color = 'mix'):\n",
    "        self.age = age\n",
    "        self.weight = weight\n",
    "        self.sex = sex\n",
    "        self.breed = breed\n",
    "        self.color = color\n",
    "        self.commands = set()\n",
    "\n",
    "    def grow(self, month = 1):\n",
    "        self.age += month\n",
    "        self.weight += month*0.5\n",
    "\n",
    "    def train(self, *commands):\n",
    "        for command in commands:\n",
    "            self.commands.add(command)\n",
    "\n",
    "    def contest_allowed(self):\n",
    "        if self.age >= 12 and self.breed != 'mongrel' and len(self.commands) >= 3:\n",
    "            return True\n",
    "        else:\n",
    "            return False"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fccef313-f791-414d-9e83-4e98921fc101",
   "metadata": {},
   "source": [
    "2. Create a class, describing RNA. It should include:\n",
    "- constructor accepts RNA sequence and creates an object with it\n",
    "- translation method\n",
    "- reverse transcription method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 354,
   "id": "c2b3975a-159f-4ef6-9c65-672d4e8a5463",
   "metadata": {},
   "outputs": [],
   "source": [
    "good_nucleotides = set('AGCUagcu')\n",
    "\n",
    "code = {\"UUU\":\"F\", \"UUC\":\"F\", \"UUA\":\"L\", \"UUG\":\"L\",\n",
    "    \"UCU\":\"S\", \"UCC\":\"s\", \"UCA\":\"S\", \"UCG\":\"S\",\n",
    "    \"UAU\":\"Y\", \"UAC\":\"Y\", \"UAA\":\"*\", \"UAG\":\"*\",\n",
    "    \"UGU\":\"C\", \"UGC\":\"C\", \"UGA\":\"*\", \"UGG\":\"W\",\n",
    "    \"CUU\":\"L\", \"CUC\":\"L\", \"CUA\":\"L\", \"CUG\":\"L\",\n",
    "    \"CCU\":\"P\", \"CCC\":\"P\", \"CCA\":\"P\", \"CCG\":\"P\",\n",
    "    \"CAU\":\"H\", \"CAC\":\"H\", \"CAA\":\"Q\", \"CAG\":\"Q\",\n",
    "    \"CGU\":\"R\", \"CGC\":\"R\", \"CGA\":\"R\", \"CGG\":\"R\",\n",
    "    \"AUU\":\"I\", \"AUC\":\"I\", \"AUA\":\"I\", \"AUG\":\"M\",\n",
    "    \"ACU\":\"T\", \"ACC\":\"T\", \"ACA\":\"T\", \"ACG\":\"T\",\n",
    "    \"AAU\":\"N\", \"AAC\":\"N\", \"AAA\":\"K\", \"AAG\":\"K\",\n",
    "    \"AGU\":\"S\", \"AGC\":\"S\", \"AGA\":\"R\", \"AGG\":\"R\",\n",
    "    \"GUU\":\"V\", \"GUC\":\"V\", \"GUA\":\"V\", \"GUG\":\"V\",\n",
    "    \"GCU\":\"A\", \"GCC\":\"A\", \"GCA\":\"A\", \"GCG\":\"A\",\n",
    "    \"GAU\":\"D\", \"GAC\":\"D\", \"GAA\":\"E\", \"GAG\":\"E\",\n",
    "    \"GGU\":\"G\", \"GGC\":\"G\", \"GGA\":\"G\", \"GGG\":\"G\",}\n",
    "\n",
    "RNA_to_DNA = {\"A\": \"T\", \"a\": \"t\", \"U\": \"A\", \"u\": \"a\",\n",
    "              \"G\": \"C\", \"g\": \"c\", \"C\": \"G\", \"c\": \"g\"}\n",
    "\n",
    "class RNA:\n",
    "    def __init__(self, sequence, protein_coding=True):\n",
    "        if type(sequence) != 'str' and set(sequence).issubset(good_nucleotides):\n",
    "            self.seq = sequence\n",
    "            self.protein_coding = protein_coding\n",
    "        else:\n",
    "            print('RNA should contain only nucleotides from set {AGCUagcu}')\n",
    "\n",
    "    def __repr__(self):\n",
    "         return self.seq\n",
    "\n",
    "    def translation(self):\n",
    "        if self.protein_coding is True:\n",
    "            codon = []\n",
    "            protein = []\n",
    "            for nucleotide in self.seq.upper():\n",
    "                codon.append(nucleotide)\n",
    "                if len(codon) == 3:\n",
    "                    protein.append(code[''.join(codon)])\n",
    "                    codon = []\n",
    "            print(''.join(protein))\n",
    "        else:\n",
    "            print('It is not a protein-coding RNA, translation is not possible')\n",
    "\n",
    "    def reverse_transcription(self):\n",
    "        dna = []\n",
    "        for nucleotide in self.seq:\n",
    "            dna.append(RNA_to_DNA[nucleotide])\n",
    "        print(''.join(dna))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31ddd3b6-9241-49cf-b7a9-f68d3de67faf",
   "metadata": {},
   "source": [
    "3. Create a class, using class set, with only positive elements."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "0688857a-15cc-4ce2-a0e4-b32b94e86ee0",
   "metadata": {},
   "outputs": [],
   "source": [
    "class PositiveSet(set):\n",
    "    def __init__(self, *elements):\n",
    "        self.set = set()\n",
    "        for element in elements:\n",
    "            if element > 0:\n",
    "                self.set.add(element)\n",
    "    def add(self, *elements):\n",
    "        for element in elements:\n",
    "            if element > 0:\n",
    "                self.set.add(element)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b2cc6278-933c-4314-b9c2-33d9028f58bf",
   "metadata": {},
   "source": [
    "4. Create a class for statistics of fasta-files, that takes a path to file in fasta format. Methods:\n",
    "- the number of sequences\n",
    "- a histogram for sequence length\n",
    "- gc-count\n",
    "- a histogram for frequences of 4-mers (x - 4mer, y - frequency)\n",
    "- change the method for print output\n",
    "- executing all methods for counting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 376,
   "id": "50bb42c4-7b3c-4317-be33-ded4336be850",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Fasta:\n",
    "    def __init__(self, path):\n",
    "        self.path = path\n",
    "\n",
    "    def __repr__(self):\n",
    "        return self.path\n",
    "        \n",
    "    def __str__(self):\n",
    "        return self.path\n",
    "\n",
    "    def count_number(self):\n",
    "        self.number = 0\n",
    "        with open(self.path) as file:\n",
    "            for line in file:\n",
    "                if line.startswith('>'):\n",
    "                    self.number += 1\n",
    "        print(self.number)\n",
    "\n",
    "    def count_gc(self):\n",
    "        self.gc = 0\n",
    "        c, g, length = 0, 0, 0\n",
    "        with open(self.path) as file:\n",
    "            for line in file:\n",
    "                if not line.startswith('>'):\n",
    "                    c += line.count('C')\n",
    "                    g += line.count('G')\n",
    "                    length += len(line)-1\n",
    "        self.gc = round((c + g)/length*100, 1)\n",
    "        print(self.gc)\n",
    "    \n",
    "    def count_all(self):\n",
    "        self.count_number()\n",
    "        self.count_gc()\n",
    "\n",
    "    def length_hist(self):\n",
    "        length_list = []\n",
    "        with open(self.path) as file:\n",
    "            length_seq = 0\n",
    "            for line in file:\n",
    "                if not line.startswith('>'):\n",
    "                    length_seq += len(line)-1\n",
    "                else:\n",
    "                    length_list.append(length_seq)\n",
    "                    length_seq = 0\n",
    "        length_list.append(length_seq)\n",
    "        length_list.remove(0)\n",
    "\n",
    "        fig, ax = plt.subplots(figsize=(15, 10))\n",
    "        plt.rcParams['font.size'] = '16'\n",
    "        for label in (ax.get_xticklabels() + ax.get_yticklabels()):\n",
    "            label.set_fontsize(16)\n",
    "        ax.hist(length_list, color=\"red\")\n",
    "        ax.set_xlabel(\"Sequence length (bp)\", fontsize=\"16\")\n",
    "        ax.tick_params(axis='y', labelcolor=\"black\")\n",
    "        ax.set_title(\"Distribution of sequence lengths over all sequences\", size=20)\n",
    "        plt.show()\n",
    "\n",
    "    def hist_4mers(self):\n",
    "        four_mers = {}\n",
    "        with open(self.path) as file:\n",
    "            for line in file:\n",
    "                if not line.startswith('>'):\n",
    "                    mer = []\n",
    "                    for nucleotide in line:\n",
    "                        if nucleotide != '\\n':\n",
    "                            mer.append(nucleotide)\n",
    "                            if len(mer) == 4:\n",
    "                                try:\n",
    "                                    four_mers[''.join(mer)] += 1\n",
    "                                except:\n",
    "                                    four_mers[''.join(mer)] = 1\n",
    "                                mer = mer[1:]\n",
    "        sum_mers = sum(four_mers.values())\n",
    "        for key, value in four_mers.items():\n",
    "            four_mers[key] = value/sum_mers\n",
    "        sorted_mers = {}\n",
    "        sorted_keys = sorted(four_mers, key=four_mers.get)\n",
    "        for key in sorted_keys:\n",
    "            sorted_mers[key] = four_mers[key]\n",
    "\n",
    "        fig, ax = plt.subplots(figsize=(70, 10))\n",
    "        for label in (ax.get_xticklabels()):\n",
    "            label.set_fontsize(10)\n",
    "        ax.bar(list(sorted_mers.keys()), sorted_mers.values(), color=\"red\")\n",
    "        ax.set_xlabel(\"4-mer frequency\", fontsize=\"16\")\n",
    "        ax.tick_params(axis='y', labelcolor=\"black\")\n",
    "        ax.set_title(\"Distribution of 4-mers\", size=20)\n",
    "        plt.xticks(rotation=90)\n",
    "        plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
